### PR TITLE
[ContainerApp] Add CLI for DotNet Components

### PR DESCRIPTION
**Pull request by snehapar9**
This checklist is used to make sure that common guidelines for a pull request are followed.
This PR adds CLI support for dotnet components. Please see video attached for the full CLI experience
https://microsoft-my.sharepoint.com/:v:/p/snehapar/ERx7M-kEDxZBnUFqMahq0FYBD8ApQYYSQa054aQsyIeR6w?e=gWsyrc
### Related command

 PR Description - This adds CLI support for to create/show/list/delete DotNet components.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [ ] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
ContainerApp

**Comment by Greedygre** [comment]
In the Integration tests failed, it is due to the vnet api-version update, you can follow to pull azure-cli dev branch latest code and setup:
```
FAILED src/containerapp/azext_containerapp/tests/latest/test_containerapp_dotnet_component.py::ContainerappDotNetComponentTests::test_containerapp_dotnet_component
FAILED src/containerapp/azext_containerapp/tests/latest/test_containerapp_env_commands.py::ContainerappEnvScenarioTest::test_containerapp_env_infrastructure_rg
FAILED src/containerapp/azext_containerapp/tests/latest/test_containerapp_nfsazurefile.py::ContainerAppMountNfsAzureFileTest::test_container_app_mount_nfsazurefile_e2e
```
But about these two tests I have rerun them in this PR: https://github.com/Azure/azure-cli-extensions/pull/7382, which will be merged after Chinese holiday.
```
FAILED src/containerapp/azext_containerapp/tests/latest/test_containerapp_env_commands.py::ContainerappEnvScenarioTest::test_containerapp_env_infrastructure_rg
FAILED src/containerapp/azext_containerapp/tests/latest/test_containerapp_nfsazurefile.py::ContainerAppMountNfsAzureFileTest::test_container_app_mount_nfsazurefile_e2e
```

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Please do the required changes

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Approved with thanks

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Please do as well

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Please do as well

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Need support

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
support pls

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Thanks

File: src/containerapp/azext_containerapp/custom.py [hunk 1]
```diff
@@ -2637,3 +2641,76 @@ def list_environment_telemetry_otlp(cmd,
...
+    raw_parameters = locals()
+    dotnet_component_decorator = DotNetComponentDecorator(
+        cmd=cmd,
+        client=DotNetComponentPreviewClient,
+        raw_parameters=raw_parameters,
+        models=CONTAINER_APPS_SDK_MODELS
+    )
+    result = dotnet_component_decorator.show()
+    if result is not None:
+        logger.warning("Found DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
```
Comment (pauld-msft) [inline src/containerapp/azext_containerapp/custom.py:2666]: why are all of these warnings? Shouldn't these be info messages?

File: src/containerapp/azext_containerapp/custom.py [hunk 2]
```diff
@@ -2637,3 +2641,76 @@ def list_environment_telemetry_otlp(cmd,
...
+    logger.warning("Deleting DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
+    result = dotnet_component_decorator.delete()
+    if result is None:
+        logger.warning("Successfully deleted DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
+    return result
+
+
+def create_dotnet_component(cmd, dotnet_component_name, environment_name, resource_group_name, no_wait, dotnet_component_type=DOTNET_COMPONENT_RESOURCE_TYPE):
+    # Check if DotNet component already exists in environment
+    existing_dotnet_component = _get_dotnet_component_if_exists(cmd, dotnet_component_name, environment_name, resource_group_name)
```
Comment (pauld-msft) [inline src/containerapp/azext_containerapp/custom.py:2690]: how would a customer update their dashboard? This structure doesn't allow for using PUT for create or update

File: src/containerapp/azext_containerapp/custom.py [hunk 3]
```diff
@@ -2637,3 +2641,76 @@ def list_environment_telemetry_otlp(cmd,
...
+    dotnet_component_decorator.construct_payload()
+    managed_environment = show_managed_environment(cmd, environment_name, resource_group_name)
+    default_domain = safe_get(managed_environment, "properties", "defaultDomain")
+    if not default_domain:
+        raise ValidationError("The containerapp environment '{}' does not have a default domain.".format(environment_name))
+    aspire_dashboard_url = f"https://{dotnet_component_name}.ext.{default_domain}"
+    logger.warning("Creating DotNet Component '%s' in environment '%s' in resource group '%s.", dotnet_component_name, environment_name, resource_group_name)
+    r = dotnet_component_decorator.create()
+    if r is not None:
+        logger.warning("Successfully created DotNet Component '%s' in environment '%s' in resource group '%s'. Access your Aspire Dashboard at %s.", dotnet_component_name, environment_name, resource_group_name, aspire_dashboard_url)
```
Comment (pauld-msft) [inline src/containerapp/azext_containerapp/custom.py:2709]: I did not see this message in your demo... I think it got hidden by the ARM payload that got returned at the end of the operation. Is there any way we could have this displayed after the payload is printed out?

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    def get_argument_component_type(self):
+        return self.get_param("dotnet_component_type")
+
+    def set_argument_component_type(self, component_type: str):
+        self.set_param("dotnet_component_type", component_type)
+
+    def construct_payload(self):
+        if self.get_argument_component_type() != "AspireDashboard":
+            logger.warning("Supported DotNet component type is: AspireDashboard. Setting component type to AspireDashboard.")
```
Comment (pauld-msft) [inline src/containerapp/azext_containerapp/dotnet_component_decorator.py:40]: Should we just override their component type like this? I feel like this could get easily missed by the customer, and it doesn't seem to add a ton of value. I'd prefer just letting our RP handle erroring out here, so we could potentially add new dotnet components in the future?

File: src/containerapp/azext_containerapp/custom.py (hunk 3 above)
Comment (pauld-msft) [inline src/containerapp/azext_containerapp/custom.py:2709]: confirm that this is an aspire dashboard `type` before printing the `Access your Aspire Dashboard at %s` part of the message

File: src/containerapp/azext_containerapp/custom.py [hunk 5]
```diff
@@ -2637,3 +2641,76 @@ def list_environment_telemetry_otlp(cmd,
...
+    if result is None:
+        logger.warning("Successfully deleted DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
+    return result
+
+
+def create_dotnet_component(cmd, dotnet_component_name, environment_name, resource_group_name, no_wait, dotnet_component_type=DOTNET_COMPONENT_RESOURCE_TYPE):
+    # Check if DotNet component already exists in environment
+    existing_dotnet_component = _get_dotnet_component_if_exists(cmd, dotnet_component_name, environment_name, resource_group_name)
+    if existing_dotnet_component:
+        raise ValidationError("DotNet Component '{}' already exists in environment '{}' in resource group '{}'. Please disable Aspire Dashboard by deleting the DotNet Component before creating a new one.".format(dotnet_component_name, environment_name, resource_group_name))
```
Comment (pauld-msft) [inline src/containerapp/azext_containerapp/custom.py:2692]: this also has `Aspire Dashboard` specific language

File: src/containerapp/azext_containerapp/_help.py [hunk 6]
```diff
@@ -1601,3 +1601,50 @@
...
+    short-summary: Command to list DotNet components within the environment.
+    examples:
+    - name: List all DotNet components within an environment.
+      text: |
+          az containerapp env dotnet-component list -g MyResourceGroup --environment MyEnvironment
+"""
+
+helps['containerapp env dotnet-component create'] = """
+    type: command
+    short-summary: Command to create DotNet component to enable Aspire Dashboard. Supported DotNet component type is Aspire Dashboard.
```
Comment (pauld-msft) [inline src/containerapp/azext_containerapp/_help.py:1621]: remove aspire dashboard specific message here, you can keep the example as-is since it shows how that type is created

**Review by pauld-msft — COMMENTED** [review]
mostly just message updates

File: src/containerapp/azext_containerapp/custom.py (hunk 1 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2666]: Info message don't show up on the console.

File: src/containerapp/azext_containerapp/custom.py (hunk 2 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2690]: We are not adding capability to update dotnet components now since we only support the component type `AspireDashboard`.

File: src/containerapp/azext_containerapp/custom.py (hunk 3 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2709]: This message is getting printed out before returning the payload. Don't think there's a good way to print this message after returning the payload to customer. 

As an alternative, we could just print this message and not return the payload at all.

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py (hunk 4 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/dotnet_component_decorator.py:40]: The goal of this is to default to `AspireDashboard` if customer provided another dotnet component type instead of erroring them out.

File: src/containerapp/azext_containerapp/_help.py (hunk 6 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/_help.py:1621]: Prefer keeping this as-is so customers know this is how they would enable an Aspire Dashboard since the dotnet-component creation command by itself does not provide too much context about how it can be used in relation to enable Aspire.

File: src/containerapp/azext_containerapp/custom.py (hunk 3 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2709]: Updated.

File: src/containerapp/azext_containerapp/custom.py (hunk 5 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2692]: Updated.

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    def get_argument_dotnet_component_name(self):
+        return self.get_param("dotnet_component_name")
+
+    def get_argument_component_type(self):
+        return self.get_param("dotnet_component_type")
+
+    def set_argument_component_type(self, component_type: str):
+        self.set_param("dotnet_component_type", component_type)
+
+    def construct_payload(self):
```
Comment (pauld-msft) [inline src/containerapp/azext_containerapp/dotnet_component_decorator.py:39]: still not super thrilled with the lack of parity to the AZD and portal experiences. Neither of those even allow the customer to name the aspire dashboard dotnet component. 

Could we maybe just add a default name of `aspire-dashboard` if the customer tries to create an dotnet component of type aspire dashboard but doesn't supply a name?

so command is a little more like a flag on the environment
`containerapp env dotnet-component create --environment my-env -g my-rg`

rather than
`containerapp env dotnet-component create --environment my-env -g my-rg -n aspire-dashboard`

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Thanks for supporting.

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py (hunk 7 above)
Comment (pauld-msft) [reply src/containerapp/azext_containerapp/dotnet_component_decorator.py:39]: spoke offline about this, keeping as is for now

**Review by pauld-msft — APPROVED** [review]

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Pls help to solve these issues.

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Thanks for supporting

**Review by pauld-msft — APPROVED** [review]

**Comment by yonzhan** [comment]
Please fix CI issues.

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Pls do the needed changes

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Thanks

File: src/containerapp/azext_containerapp/_params.py [hunk 8]
```diff
@@ -344,3 +344,9 @@ def load_arguments(self, _):
         c.argument('environment_name', options_list=['--environment'], help="The environment name.")
         c.argument('resource_group_name', arg_type=resource_group_name_type, id_part=None)
         c.argument('configuration', nargs="*", help="Java component configuration. Configuration must be in format \"<propertyName>=<value> <propertyName>=<value> ...\".")
+
+    with self.argument_context('containerapp env dotnet-component') as c:
+        c.argument('dotnet_component_name', options_list=['--name', '-n'], help="The DotNet component name.")
+        c.argument('environment_name', options_list=['--environment'], help="The environment name.")
+        c.argument('resource_group_name', arg_type=resource_group_name_type, id_part=None)
+        c.argument('dotnet_component_type', options_list = ['--type'], help="The type of DotNet component. Supported type is 'AspireDashboard'.")
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_params.py:352]: For supported type:
`, arg_type=get_enum_type(['AspireDashboard'])`

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        if self.get_argument_component_type() == "":
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/dotnet_component_decorator.py:40]: If you using `arg_type=get_enum_type(['AspireDashboard'])` in the params, not need this check.

File: src/containerapp/azext_containerapp/custom.py [hunk 10]
```diff
@@ -2638,3 +2642,87 @@ def list_environment_telemetry_otlp(cmd,
...
+        models=CONTAINER_APPS_SDK_MODELS
+    )
+    result = dotnet_component_decorator.show()
+    if result is not None:
+        logger.warning("Found DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
+
+    aspire_dashboard_url = _get_aspire_dashboard_url(cmd, environment_name, resource_group_name, dotnet_component_name)
+    component_type = safe_get(result, "properties", "componentType")
+    if component_type == DOTNET_COMPONENT_RESOURCE_TYPE:
+        logger.warning("Aspire Dashboard URL: %s.", aspire_dashboard_url)
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/custom.py:2672]: The logic about gets `aspire_dashboard_url` should be moved tunder `if component_type == DOTNET_COMPONENT_RESOURCE_TYPE:`, right?
```
    aspire_dashboard_url = _get_aspire_dashboard_url(cmd, environment_name, resource_group_name, dotnet_component_name)
    component_type = safe_get(result, "properties", "componentType")
```

File: src/containerapp/azext_containerapp/custom.py [hunk 11]
```diff
@@ -2638,3 +2642,87 @@ def list_environment_telemetry_otlp(cmd,
...
+        raise ValidationError("DotNet Component '{}' does not exist in environment '{}' in resource group '{}'.".format(dotnet_component_name, environment_name, resource_group_name))
+    raw_parameters = locals()
+    dotnet_component_decorator = DotNetComponentDecorator(
+        cmd=cmd,
+        client=DotNetComponentPreviewClient,
+        raw_parameters=raw_parameters,
+        models=CONTAINER_APPS_SDK_MODELS
+    )
+    logger.warning("Deleting DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
+    result = dotnet_component_decorator.delete()
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/custom.py:2688]: `dotnet_component_decorator.delete()` return nothing, why need to check if the `result` is `None` here?

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        self.dotnet_component_def["properties"]["componentType"] = self.get_argument_component_type()
+
+    def create(self):
+        try:
+            return self.client.create(
+                cmd=self.cmd, resource_group_name=self.get_argument_resource_group_name(),
+                environment_name=self.get_argument_environment_name(), name=self.get_argument_dotnet_component_name(),
+                dotnet_component_envelope=self.dotnet_component_def, no_wait=self.get_argument_no_wait())
+        except Exception as e:
+            stringErr = str(e)
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/dotnet_component_decorator.py:52]: `stringErr` -> `string_err`

File: src/containerapp/azext_containerapp/custom.py [hunk 13]
```diff
@@ -2638,3 +2642,87 @@ def list_environment_telemetry_otlp(cmd,
...
+        return show_dotnet_component(cmd, dotnet_component_name, environment_name, resource_group_name)
+    except Exception as e:
+        return None
+
+def _get_aspire_dashboard_url(cmd, environment_name, resource_group_name, dotnet_component_name):
+    managed_environment = show_managed_environment(cmd, environment_name, resource_group_name)
+    default_domain = safe_get(managed_environment, "properties", "defaultDomain")
+    if not default_domain:
+        raise ValidationError("The containerapp environment '{}' does not have a default domain.".format(environment_name))
+    return f"https://{dotnet_component_name}.ext.{default_domain}"
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/custom.py:2728]: Recommend moving this to `_util.py`

File: src/containerapp/azext_containerapp/custom.py [hunk 14]
```diff
@@ -2638,3 +2642,87 @@ def list_environment_telemetry_otlp(cmd,
...
+    if result is None:
+        logger.warning("Successfully deleted DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
+    return result
+
+
+def create_dotnet_component(cmd, dotnet_component_name, environment_name, resource_group_name, no_wait, dotnet_component_type=DOTNET_COMPONENT_RESOURCE_TYPE):
+    # Check if DotNet component already exists in environment
+    existing_dotnet_component = _get_dotnet_component_if_exists(cmd, dotnet_component_name, environment_name, resource_group_name)
+    if existing_dotnet_component:
+        raise ValidationError("DotNet Component '{}' already exists in environment '{}' in resource group '{}'. Please delete the DotNet Component before creating a new one.".format(dotnet_component_name, environment_name, resource_group_name))
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/custom.py:2698]: Recommend moving this validation logic to `DotNetComponentCreateDecorator.validate_arguments()`

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py (hunk 9 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/dotnet_component_decorator.py:40]: The expected behavior is we need to default the component type to `AspireDashboard` if customer did not provide a `--type` argument. so this check is still required?

File: src/containerapp/azext_containerapp/custom.py [hunk 15]
```diff
@@ -2638,3 +2642,87 @@ def list_environment_telemetry_otlp(cmd,
...
+        client=DotNetComponentPreviewClient,
+        raw_parameters=raw_parameters,
+        models=CONTAINER_APPS_SDK_MODELS
+    )
+    dotnet_component_decorator.construct_payload()
+    aspire_dashboard_url = _get_aspire_dashboard_url(cmd, environment_name, resource_group_name, dotnet_component_name)
+    logger.warning("Creating DotNet Component '%s' in environment '%s' in resource group '%s.", dotnet_component_name, environment_name, resource_group_name)
+    r = dotnet_component_decorator.create()
+    if r is not None:
+        logger.warning("Successfully created DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
```
[the remark this answers is not in the document]
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2711]: Yes, they are required so they show up on the customer's console.

File: src/containerapp/azext_containerapp/custom.py (hunk 10 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2672]: Updated.

File: src/containerapp/azext_containerapp/custom.py (hunk 11 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2688]: okay, thanks! looks like an exception will be thrown if deletion errored out. So I'm removing the check for result to be None here.

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py (hunk 12 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/dotnet_component_decorator.py:52]: Updated.

File: src/containerapp/azext_containerapp/custom.py (hunk 13 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2728]: Prefer keeping it here since this method makes a call to `show_managed_environment` in custom.py

Comment (Greedygre) [reply src/containerapp/azext_containerapp/custom.py:2728]: About `call show_managed_environment`, it is easy to replace it with:
```
        try:
            env_info = ManagedEnvironmentPreviewClient.show(cmd=self.cmd, resource_group_name=env_rg, name=env_name)
        except Exception as e:
            handle_non_404_status_code_exception(e)
```

File: src/containerapp/azext_containerapp/_params.py (hunk 8 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/_params.py:352]: Updated.

File: src/containerapp/azext_containerapp/custom.py (hunk 14 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2698]: Updated.

File: src/containerapp/azext_containerapp/_params.py (hunk 8 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/_params.py:352]: Expectation is the following image. This is already being handled by CP
image

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py (hunk 9 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/dotnet_component_decorator.py:40]: Keeping this as-is because of [this response](https://github.com/Azure/azure-cli-extensions/pull/7564#discussion_r1597921969)

File: src/containerapp/azext_containerapp/_params.py (hunk 8 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/_params.py:352]: image
This is wired and not expected. Can you show me the code of the params you changed?

Comment (Greedygre) [reply src/containerapp/azext_containerapp/_params.py:352]: @snehapar9 
 
It is a common recommend setting in CLI, and I tested:
image
It works fine:
`        c.argument('dotnet_component_type', options_list=['--type'], help="The type of DotNet component. Supported type is 'AspireDashboard'.", arg_type=get_enum_type(['AspireDashboard']))`

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_dotnet_component.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+        env_create_cmd = f'containerapp env create -g {resource_group} -n {env_name} --location {location} --logs-destination none --enable-workload-profiles'
+        self.cmd(env_create_cmd)
+
+        # List DotNet Components
+        dotnet_component_list = self.cmd("containerapp env dotnet-component list -g {} --environment {}".format(resource_group, env_name)).get_output_in_json()
+        self.assertTrue(len(dotnet_component_list) == 0)
+
+        # Create DotNet Component
+        self.cmd('containerapp env dotnet-component create -g {} -n {} --environment {}'.format(resource_group, dotnet_component_name, env_name))
```
[the remark this answers is not in the document]
Comment (snehapar9) [reply src/containerapp/azext_containerapp/tests/latest/test_containerapp_dotnet_component.py:29]: Updated.

File: src/containerapp/azext_containerapp/_params.py (hunk 8 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/_params.py:352]: image

Comment (Greedygre) [inline src/containerapp/azext_containerapp/_params.py:352]:
```suggestion
        c.argument('dotnet_component_type', options_list=['--type'], help="The type of DotNet component.", arg_type=get_enum_type(['AspireDashboard']))

```

Comment (snehapar9) [reply src/containerapp/azext_containerapp/_params.py]: Okay, got it. I've updated it.

File: src/containerapp/azext_containerapp/custom.py [hunk 17]
```diff
@@ -2638,3 +2642,76 @@ def list_environment_telemetry_otlp(cmd,
...
+    component_type = safe_get(r, "properties", "componentType")
+    if component_type == DOTNET_COMPONENT_RESOURCE_TYPE:
+        logger.warning("Access your Aspire Dashboard at %s.", aspire_dashboard_url)
+    return
+
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/custom.py:2711]: Need one more blank line:
image

File: src/containerapp/azext_containerapp/custom.py [hunk 18]
```diff
@@ -2638,3 +2642,76 @@ def list_environment_telemetry_otlp(cmd,
...
+        logger.warning("Found DotNet Component '%s' in environment '%s' in resource group '%s'.", dotnet_component_name, environment_name, resource_group_name)
+
+    component_type = safe_get(result, "properties", "componentType")
+    if component_type == DOTNET_COMPONENT_RESOURCE_TYPE:
+        aspire_dashboard_url = _get_aspire_dashboard_url(cmd, environment_name, resource_group_name, dotnet_component_name)
+        logger.warning("Aspire Dashboard URL: %s.", aspire_dashboard_url)
+    return result
+
+
+def delete_dotnet_component(cmd, dotnet_component_name, environment_name, resource_group_name, no_wait):
```
[the remark this answers is not in the document]
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py:2676]: @Greedygre this covers it - https://github.com/Azure/azure-cli-extensions/pull/7564/files#diff-0265fc6cc0ea8af5a57382fcd906bac145060abb23953c9a300ea0ed19e93f4bR100

[the remark this answers is not in the document]
Comment (Greedygre) [reply src/containerapp/azext_containerapp/custom.py:2676]: About no-wait, it should have a default value `False` as other command's `no-wait`:

File: src/containerapp/azext_containerapp/_params.py [hunk 19]
```diff
@@ -344,3 +344,9 @@ def load_arguments(self, _):
         c.argument('environment_name', options_list=['--environment'], help="The environment name.")
         c.argument('resource_group_name', arg_type=resource_group_name_type, id_part=None)
         c.argument('configuration', nargs="*", help="Java component configuration. Configuration must be in format \"<propertyName>=<value> <propertyName>=<value> ...\".")
+
+    with self.argument_context('containerapp env dotnet-component') as c:
+        c.argument('dotnet_component_name', options_list=['--name', '-n'], help="The DotNet component name.")
+        c.argument('environment_name', options_list=['--environment'], help="The environment name.")
+        c.argument('resource_group_name', arg_type=resource_group_name_type, id_part=None)
+        c.argument('dotnet_component_type', options_list = ['--type'], arg_type=get_enum_type(['AspireDashboard']), help="The type of DotNet component. Supported type is 'AspireDashboard'.")
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_params.py:352]: As you already has  `arg_type=get_enum_type(['AspireDashboard']`, not need `Supported type is 'AspireDashboard'` in the message.

Comment (snehapar9) [reply src/containerapp/azext_containerapp/_params.py:352]: Updated.

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+from ._client_factory import handle_raw_exception
+from ._clients import ManagedEnvironmentPreviewClient
+
+logger = get_logger(__name__)
+
+
+class DotNetComponentDecorator(BaseResource):
+    def __init__(self, cmd: AzCliCommand, client: Any, raw_parameters: Dict, models: str):
+        super().__init__(cmd, client, raw_parameters, models)
+        self.dotnet_component_def = DotNetComponentModel
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/dotnet_component_decorator.py:27]:
```suggestion
        self.dotnet_component_def = deepcopy(DotNetComponentModel)
```

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+            return None
+
+    def validate_arguments(self, dotnet_component_name, environment_name, resource_group_name, validation_error):
+        # Check if DotNet component already exists in environment
+        existing_dotnet_component = self._get_dotnet_component_if_exists(dotnet_component_name, environment_name, resource_group_name)
+        if existing_dotnet_component:
+            raise ValidationError(validation_error)
+
+    def _get_aspire_dashboard_url(self, environment_name, resource_group_name, dotnet_component_name):
+        managed_environment = ManagedEnvironmentPreviewClient.show(self.cmd, environment_name, resource_group_name)
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/dotnet_component_decorator.py:96]:
```suggestion
        managed_environment = ManagedEnvironmentPreviewClient.show(self.cmd, resource_group_name, environment_name)
```

File: src/containerapp/azext_containerapp/custom.py (hunk 13 above)
Comment (snehapar9) [reply src/containerapp/azext_containerapp/custom.py]: Updated.

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Pls do as required

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Thanks

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_dotnet_component.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+
+        env_create_cmd = f'containerapp env create -g {resource_group} -n {env_name} --location {location} --logs-destination none --enable-workload-profiles'
+        self.cmd(env_create_cmd)
+
+        # List DotNet Components
+        dotnet_component_list = self.cmd("containerapp env dotnet-component list -g {} --environment {}".format(resource_group, env_name)).get_output_in_json()
+        self.assertTrue(len(dotnet_component_list) == 0)
+
+        # Create DotNet Component
+        self.cmd('containerapp env dotnet-component create -g {} -n {} --environment {} --type {}'.format(resource_group, dotnet_component_name, env_name, "test-component-type"), expect_failure=True)
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/tests/latest/test_containerapp_dotnet_component.py:64]: In CLI it will throw SystemExit 2 for `Parser error; check input to command line`.
```
with self.assertRaisesRegex(SystemExit, '2'):
    self.cmd('containerapp env dotnet-component create -g a -n b --environment c --type {}'.format("test-component-type"))
```

Comment (snehapar9) [reply src/containerapp/azext_containerapp/tests/latest/test_containerapp_dotnet_component.py]: Thanks!

**Review by Greedygre — APPROVED** [review]

File: src/containerapp/azext_containerapp/dotnet_component_decorator.py [hunk 23]
```diff
@@ -82,17 +82,25 @@ def delete(self):
...
         existing_dotnet_component = self._get_dotnet_component_if_exists(dotnet_component_name, environment_name, resource_group_name)
         if existing_dotnet_component:
             raise ValidationError(validation_error)
 
+    def validate_delete_arguments(self, dotnet_component_name, environment_name, resource_group_name, validation_error):
+        # Check if DotNet component exists in environment
+        existing_dotnet_component = self._get_dotnet_component_if_exists(dotnet_component_name, environment_name, resource_group_name)
+        if not existing_dotnet_component:
+            raise ValidationError(validation_error)
+
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/dotnet_component_decorator.py:103]: Recommend removing them both, sync offline, as this has been handle by CP.

Comment (snehapar9) [reply src/containerapp/azext_containerapp/dotnet_component_decorator.py]: Removed. Thanks

**Review by Greedygre — APPROVED** [review]

File: src/containerapp/azext_containerapp/_help.py [hunk 24]
```diff
@@ -1684,3 +1684,132 @@
...
 
+# SessionPool Commands
+helps['containerapp sessionpool'] = """
+    type: group
+    short-summary: Commands to manage session pools.
+"""
+
+helps['containerapp sessionpool create'] = """
+    type: command
+    short-summary: Create or update a Session pool.
```
Comment (yanzhudd) [inline src/containerapp/azext_containerapp/_help.py:1695]: In CLI, create command is only used for create operation
```suggestion
    short-summary: Create a Session pool.
```

Comment (yanzhudd) [reply src/containerapp/azext_containerapp/_help.py:1695]: please check the similar messages as well.

Comment (Greedygre) [reply src/containerapp/azext_containerapp/_help.py:1695]: Hi @yanzhudd 

This command support update too.

File: src/containerapp/azext_containerapp/_help.py [hunk 25]
```diff
@@ -1684,3 +1684,132 @@
           az containerapp sessionpool list -g MyResourceGroup
 """
 
+# SessionPool Commands
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_help.py:1687]: Hi @snehapar9 
This is not related to your change. Please check

File: src/containerapp/azext_containerapp/_help.py [hunk 26]
```diff
@@ -1684,3 +1684,132 @@
...
+    type: command
+    short-summary: List Session Pools by subscription or resource group.
+    examples:
+    - name: List Session Pools in the current subscription.
+      text: |
+          az containerapp sessionpool list
+    - name: List Session Pools by resource group.
+      text: |
+          az containerapp sessionpool list -g MyResourceGroup
+"""
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_help.py:1768]: @snehapar9 
This is not related to your change. Please check

**Review by yanzhudd — APPROVED** [review]

**Review by Moazzem-Hossain-pixel — COMMENTED** [review]
Need support

**Review by Moazzem-Hossain-pixel — COMMENTED** [review]
Need support
